pkg/logs/input/container/launcher.go

```diff
@@ -6,80 +6,109 @@
-// a kubernetes launcher if '/var/log/pods' is mounted ; this behaviour is reversed when
-// kubernetesCollectFromFiles is enabled.
-// If dockerCollectFromFiles is enabled the docker launcher will first attempt to tail
-// containers from file instead of the docker socket if '/var/lib/docker/containers'
-// is mounted.
-// If none of those volumes are mounted, returns a lazy docker launcher with a retrier to handle the cases
-// where docker is started after the agent.
-// dockerReadTimeout is a configurable read timeout for the docker client.
-func NewLauncher(collectAll bool,
-	kubernetesCollectFromFiles bool,
-	dockerCollectFromFiles bool,
-	dockerForceCollectFromFile bool,
-	dockerReadTimeout time.Duration,
-	sources *config.LogSources,
-	services *service.Services,
-	pipelineProvider pipeline.Provider,
-	registry auditor.Registry) restart.Restartable {
-	var (
-		launcher restart.Restartable
-		err      error
-	)
+// Launchable is a retryable wrapper for a restartable
+type Launchable struct {
+	IsAvailable func() (bool, *retry.Retrier)
+	Launcher    func() restart.Restartable
+}
 
-	if kubernetesCollectFromFiles {
-		launcher, err = kubernetes.NewLauncher(sources, services, collectAll)
-		if err == nil {
-			log.Info("Kubernetes launcher initialized")
-			return launcher
-		}
-		log.Infof("Could not setup the kubernetes launcher: %v", err)
+// Launcher tries to select a container launcher and retry on failure
+type Launcher struct {
+	containerLaunchables []Launchable
+	activeLauncher       restart.Restartable
+	stop                 bool
+	lock                 sync.Mutex
+}
 
-		launcher, err = docker.NewLauncher(dockerReadTimeout, sources, services, pipelineProvider, registry, false, dockerCollectFromFiles, dockerForceCollectFromFile)
-		if err == nil {
-			log.Info("Docker launcher initialized")
-			return launcher
-		}
-		log.Infof("Could not setup the docker launcher: %v", err)
-	} else {
-		launcher, err = docker.NewLauncher(dockerReadTimeout, sources, services, pipelineProvider, registry, false, dockerCollectFromFiles, dockerForceCollectFromFile)
-		if err == nil {
-			log.Info("Docker launcher initialized")
-			return launcher
-		}
-		log.Infof("Could not setup the docker launcher: %v", err)
+// NewLauncher creates a new launcher
+func NewLauncher(containerLaunchers []Launchable) *Launcher {
+	return &Launcher{
+		containerLaunchables: containerLaunchers,
+	}
+}
 
-		launcher, err = kubernetes.NewLauncher(sources, services, collectAll)
-		if err == nil {
-			log.Info("Kubernetes launcher initialized")
-			return launcher
+func (l *Launcher) launch(launchable Launchable) {
+	launcher := launchable.Launcher()
+	if launcher == nil {
+		launcher = NewNoopLauncher()
+	}
+	l.activeLauncher = launcher
+	l.activeLauncher.Start()
+}
+
+func (l *Launcher) shouldRetry() (bool, time.Duration) {
+	var retryer *retry.Retrier
+	for _, launchable := range l.containerLaunchables {
```

Comment:
A question I have is: order is important, but it only seems important in the context of each launch attempt? So we have k8s, docker launchables.... in each attempt we look first at k8s, but if for whatever reason the `docker` launchable is available first, we'll use that one. I guess what I'm trying to understand is if it's OK if docker wins the race to being `launchable.IsAvailable() == true` even though the k8s one might be available shortly after.

Reply:
Yes - this is the case. To give a little bit more background: The issue you described will only happen if they are using both docker and K8s and K8s fails first. Part of this change is to address users who use a different container backend. Today if K8s fails and they are not using docker, we still only spawn a lazy docker launcher which will never be used. 

With this change docker will always fail in that case and as a result - keep retrying the kubelet. 

@olivielpeau (summarizing your description from the source card) - but tagging you in case you have more comments on this.

Reply:
Correct 👍 

Given that the existing behavior was to only try each launcher once, indeed keeping the order only in the context of each launch attempt is OK (since it shouldn't introduce regressions compared to the existing behavior).

Or, to put it in your own words, it's OK if docker wins the race to being `launchable.IsAvailable() == true` even though the k8s one might be available shortly after, because this behavior (although not ideal) is not a regression over the existing behavior. It would be even better if in this scenario the Logs Agent could revert to using the k8s launcher once it's available, but I believe that would be a much more difficult change to make.

Reply:
I'm good with keeping this, just wanted to make sure this was all intentional.

Reply:
Works for us 👍